src/Razor/src/Microsoft.CodeAnalysis.Razor.CohostingShared/CohostStartupService.cs

```diff
@@ -5,45 +5,43 @@
 using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.ComponentModel.Composition;
+using System.Linq;
 using System.Text.Json;
 using System.Threading;
 using System.Threading.Tasks;
 using Microsoft.CodeAnalysis.ExternalAccess.Razor.Cohost;
+using Microsoft.CodeAnalysis.Razor.Logging;
 using Microsoft.CodeAnalysis.Razor.Protocol;
-using Microsoft.CodeAnalysis.Razor.Workspaces;
 
 namespace Microsoft.VisualStudio.Razor.LanguageClient.Cohost;
 
 [Export(typeof(ICohostStartupService))]
 [method: ImportingConstructor]
 internal class CohostStartupService(
-    LanguageServerFeatureOptions languageServerFeatureOptions,
-    [ImportMany] IEnumerable<Lazy<IRazorCohostStartupService>> lazyStartupServices)
-    : ICohostStartupService
+    [ImportMany] IEnumerable<Lazy<IRazorCohostStartupService>> lazyStartupServices,
+    ILoggerFactory loggerFactory) : ICohostStartupService
 {
-    private readonly LanguageServerFeatureOptions _languageServerFeatureOptions = languageServerFeatureOptions;
     private readonly ImmutableArray<Lazy<IRazorCohostStartupService>> _lazyStartupServices = [.. lazyStartupServices];
+    private readonly ILogger _logger = loggerFactory.GetOrCreateLogger<CohostStartupService>();
 
     public async Task StartupAsync(string clientCapabilitiesString, RazorCohostRequestContext requestContext, CancellationToken cancellationToken)
     {
         var clientCapabilities = JsonSerializer.Deserialize<VSInternalClientCapabilities>(clientCapabilitiesString, JsonHelpers.JsonSerializerOptions) ?? new();
 
-        // We want to make sure LanguageServerFeatureOptions is initialized first, so decisions can be made based on flags,
-        // and we also ensure we have a RazorClientLanguageServerManager available for anything that needs to call back to the client,
-        // so we do those two things first, manually.
-        if (_languageServerFeatureOptions is IRazorCohostStartupService startupService)
-        {
-            await startupService.StartupAsync(clientCapabilities, requestContext, cancellationToken).ConfigureAwait(false);
-        }
+        var providers = _lazyStartupServices.Select(p => p.Value).OrderByAsArray(p => p.Order);
 
-        if (!_languageServerFeatureOptions.UseRazorCohostServer)
+        foreach (var provider in providers)
         {
-            return;
+            try
+            {
+                await provider.StartupAsync(clientCapabilities, requestContext, cancellationToken).ConfigureAwait(false);
+            }
+            catch (Exception ex)
```

Comment:
Should we handle cancellation? What happens if `cancellationToken` is actually cancelled? I'm guessing that we'll be in a partially loaded state.

Reply:
I would imagine cancellation here only happens if the server is being killed, but its a good call to add something in.